src/lib/crypto/kimchi_backend/common/plonk_constraint_system.ml

```diff
@@ -268,13 +254,15 @@ type 'f t =
     mutable rows_rev : V.t option array list
   ; (* A circuit is described by a series of gates.
        A gate is finalized once [finalize_and_get_gates] is called.
+       The finalized tag contains the digest of the circuit.
     *)
     mutable gates :
-      [ `Finalized | `Unfinalized_rev of (unit, 'f) Gate_spec.t list ]
+      [ `Finalized of
+        Core.Md5.t * 'rust_gates
+        (* TODO: why is this a polymorphic variant? *)
```

Comment:
that's the big change. 

* before: `Finalized` contained nothing
* after: `Finalized` contains the digest, and the compiled circuit (a Rust vector of gates)

Reply:
Also if someone can answer the question in the TODO so I can remove it :)

Reply:
I guess it's a polymorphic variant because I didn't feel like defining a type. (It's like using a tuple instead of a record.) Feel free to add one though!

Reply:
Thanks for the answer! I remember reading that polymorphic variants had a cost. Can’t remember what though :)

Reply:
Thread on that: https://discuss.ocaml.org/t/polymorphic-variant-representation/5904